File: sdk/kotlin/examples/src/main/kotlin/app/rocksky/examples/Remote.kt
/**
 * Remote player + controller tour (needs a real Rocksky access token).
 *
 *   ./gradlew :examples:run -PmainClass=app.rocksky.examples.RemoteKt --args="<ACCESS_TOKEN>"
 *
 * A RemotePlayer advertises what "this device" is playing and reacts to commands
 * from a controller; a RemoteController lists the user's players, picks a
 * primary, observes their state, and sends commands. Both use a background poll
 * thread — configure handlers in `listen { … }`.
 *
 * Needs the native lib — run ../build-core.sh first.
 */
package app.rocksky.examples

import app.rocksky.RemoteNowPlaying
import app.rocksky.RemoteQueueItem
import app.rocksky.RemoteStatus
import app.rocksky.listen
import app.rocksky.remoteController
import app.rocksky.remotePlayer

fun main(args: Array<String>) {
    val token = args.firstOrNull() ?: run {
        println("usage: RemoteKt <ACCESS_TOKEN>")
        return
    }

    // ── player: expose this process as a controllable device ─────────────────
    val player = remotePlayer(token, "Kotlin Example Player")
    player.listen {
        onPlay = { println("[player] play") }
        onPause = { println("[player] pause") }
        onNext = { println("[player] next") }
        onPrevious = { println("[player] previous") }
        onSeek = { ms -> println("[player] seek -> ${ms}ms") }
        onQueueJump = { i -> println("[player] queue jump -> $i") }
        onQueueRemove = { i -> println("[player] queue remove -> $i") }
        onEnqueue = { cmd -> println("[player] enqueue ${cmd.tracks.size} track(s), mode=${cmd.mode}") }
    }

    // Advertise a track + queue so controllers have something to show.
    player.setNowPlaying(
        RemoteNowPlaying(
            title = "Chaser",
            artist = "Calibro 35",
            album = "Jazzploitation",
            albumArtist = "Calibro 35",
            albumArt = "",
            durationMs = 182320uL,
            elapsedMs = 0uL,
            isPlaying = true,
        ),
    )
    player.setStatus(RemoteStatus.PLAYING)
    player.setQueue(
        listOf(
            RemoteQueueItem(
                uploadId = "",
                trackId = "",
                title = "Chaser",
                artist = "Calibro 35",
                album = "Jazzploitation",
                albumArtist = "Calibro 35",
                albumArt = "",
                durationMs = 182320uL,
                songUri = "",
                albumUri = "",
                trackNumber = 0,
            ),
        ),
        0u,
    )

    // ── controller: observe + drive the user's players ───────────────────────
    val controller = remoteController(token, "Kotlin Example Controller")
    controller.listen {
        onDevices = { e -> println("[controller] ${e.devices.size} device(s), primary=${e.primaryDevice}") }
        onNowPlaying = { e -> println("[controller] ${e.deviceName}: ${e.track.artist} — ${e.track.title}") }
        onStatus = { e -> println("[controller] ${e.deviceName}: ${e.status}") }
        onQueue = { e -> println("[controller] ${e.deviceName}: queue index ${e.index}") }
    }

    // Drive whatever devices exist (null target = broadcast to all).
    Thread.sleep(1000)
    controller.pause(null)
    controller.seek(null, 42000uL)

    println("listening for 30s — control this player from the Rocksky miniplayer…")
    Thread.sleep(30_000)

    controller.disconnect()
    player.disconnect()
}

File: sdk/typescript/examples/remote.ts
// Remote control over the WebSocket: a controllable player and a controller.
//   ROCKSKY_TOKEN=<jwt> bun run examples/remote.ts player
//   ROCKSKY_TOKEN=<jwt> bun run examples/remote.ts controller
import { RemoteController, RemotePlayer } from "../src/index.js";

const token = process.env.ROCKSKY_TOKEN;
if (!token) throw new Error("set ROCKSKY_TOKEN to a Rocksky access token");
const role = process.argv[2] ?? "player";

if (role === "controller") {
  const controller = new RemoteController({ token, name: "TS Example Controller" });
  controller
    .on("devices", ({ primaryDevice, devices }) => {
      console.log(`devices: ${devices.length} (primary ${primaryDevice ?? "none"})`);
      for (const d of devices) {
        const np = d.nowPlaying ? `${d.nowPlaying.artist} — ${d.nowPlaying.title}` : "(idle)";
        console.log(`  · ${d.name} [${d.deviceId}] ${np}`);
      }
      // Make the first device primary and pause it, as a demo.
      if (devices[0]) {
        controller.setPrimary(devices[0].deviceId);
        controller.pause(devices[0].deviceId);
      }
    })
    .on("nowPlaying", ({ deviceName, track }) =>
      console.log(`♪ ${deviceName}: ${track.artist} — ${track.title}`),
    )
    .on("status", ({ deviceName, status }) => console.log(`● ${deviceName}: ${status}`))
    .on("primaryChanged", ({ deviceId }) => console.log(`★ primary → ${deviceId}`));
  controller.connect();
  console.log("controller connected — listening (Ctrl-C to quit)…");
} else {
  const player = new RemotePlayer({ token, name: "TS Example Player" });
  player
    .on("play", () => console.log("▶ play"))
    .on("pause", () => console.log("⏸ pause"))
    .on("next", () => console.log("⏭ next"))
    .on("previous", () => console.log("⏮ previous"))
    .on("seek", (ms) => console.log(`⏩ seek ${ms}ms`))
    .on("enqueue", (cmd) => console.log(`+ enqueue ${cmd.tracks.length} track(s) (${cmd.mode})`));
  player.connect();

  player.setNowPlaying({
    title: "Chaser",
    artist: "Calibro 35",
    album: "Jazzploitation",
    durationMs: 182320,
    isPlaying: true,
  });
  player.setStatus("playing");
  console.log("player connected — waiting for commands (Ctrl-C to quit)…");
}
